src/lib/annotations/AnnotationController.js

```diff
@@ -0,0 +1,146 @@
+     *
+     * @public
+     * @param {Annotator} annotator - The annotator to be associated with the controller
+     * @return {void}
+     */
+    registerAnnotator(annotator) {
+        // TODO (@minhnguyen): remove the need to register an annotator. Ideally, the annotator should know about the
+        //                     controller and the controller does not know about the annotator.
+        this.annotator = annotator;
+    }
+
+    /**
+     * Bind the mode listeners and store each handler for future unbinding
+     *
+     * @public
+     * @return {void}
+     */
+    bindModeListeners() {
+        const handlers = this.setupAndGetHandlers();
+        handlers.forEach((handler) => {
+            const eventNames = handler.type.split(' ');
+            eventNames.forEach((eventName) => handler.eventObj.addEventListener(eventName, handler.func));
+            this.handlers.push(handler);
+        });
+    }
+
+    /**
+     * Unbind the previously bound mode listeners
+     *
+     * @public
+     * @return {void}
+     */
+    unbindModeListeners() {
+        while (this.handlers.length > 0) {
+            const handler = this.handlers.pop();
+            const eventNames = handler.type.split(' ');
+            eventNames.forEach((eventName) => {
+                handler.eventObj.removeEventListener(eventName, handler.func);
+            });
+        }
+    }
+
+    /**
+     * Register a thread with the controller so that the controller can keep track of relevant threads
+     *
+     * @public
+     * @param {AnnotationThread} thread - The thread to register with the controller
+     * @return {void}
+     */
+    registerThread(thread) {
+        this.threads.push(thread);
+    }
+
+    /**
+     * Unregister a previously registered thread
+     *
+     * @public
+     * @param {AnnotationThread} thread - The thread to unregister with the controller
+     * @return {void}
+     */
+    unregisterThread(thread) {
+        this.threads = this.threads.filter((item) => item !== thread);
+    }
+
+    /**
+     * Binds custom event listeners for a thread.
+     *
+     * @protected
+     * @param {AnnotationThread} thread - Thread to bind events to
+     * @return {void}
+     */
+    bindCustomListenersOnThread(thread) {
+        if (!thread) {
+            return;
+        }
+
+        // TODO (@minhnguyen): Move annotator.bindCustomListenersOnThread logic to AnnotationController
+        this.annotator.bindCustomListenersOnThread(thread);
+        thread.addListener('annotationevent', (data) => {
+            this.handleAnnotationEvent(thread, data);
+        });
+    }
+
+    /**
+     * Unbinds custom event listeners for the thread.
+     *
+     * @protected
+     * @param {AnnotationThread} thread - Thread to unbind events from
+     * @return {void}
+     */
+    unbindCustomListenersOnThread(thread) {
+        if (!thread) {
+            return;
+        }
+
+        thread.removeAllListeners('threaddeleted');
```

Comment:
Looks like you're right about this, but there is a comment in annotator regarding calling unbindCustomListenersOnThread on 'threadcleanup':
>// Thread should be cleaned up, unbind listeners - we don't do 
>// in threaddeleted listener since thread may still need to respond
>// to error messages